app/src/block_hash.rs

```diff
@@ -1,37 +1,226 @@
+        }
+    }
+}
+
+impl Default for SharedBlockHashCache {
+    fn default() -> Self {
+        Self {
+            inner: Arc::new(RwLock::new(BlockHashRingBuffer::default())),
+        }
+    }
+}
+
+impl BlockHashLookup for SharedBlockHashCache {
     fn hash_by_number(&self, number: u64) -> Option<B256> {
-        let block = self
-            .block_query
-            .by_height(self.storage, number, false)
-            .ok()??;
-        Some(block.0.header.hash)
+        if let Ok(cache) = self.inner.read() {
+            cache.hash_by_number(number)
+        } else {
+            // TODO: deal with poisoning?
+            None
+        }
+    }
+}
+
+impl BlockHashWriter for SharedBlockHashCache {
+    fn push(&mut self, height: u64, hash: B256) {
+        if let Ok(mut cache) = self.inner.write() {
+            cache.push(height, hash)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_ring_buffer_basic_operations() {
+        let mut buffer = BlockHashRingBuffer::default();
+        assert!(buffer.entries.is_empty());
+
+        let hash1 = B256::from([1u8; 32]);
+        let hash2 = B256::from([2u8; 32]);
+
+        buffer.push(1, hash1);
+        buffer.push(2, hash2);
+
+        assert_eq!(buffer.entries.len(), 2);
+        assert_eq!(buffer.hash_by_number(1), Some(hash1));
+        assert_eq!(buffer.hash_by_number(2), Some(hash2));
+        assert_eq!(buffer.latest_block, 2);
+    }
+
+    #[test]
+    fn test_ring_buffer_capacity() {
+        let mut buffer = BlockHashRingBuffer::default();
+
+        // Fill beyond capacity
+        for i in 0..300u64 {
+            let hash = B256::from([(i % 256) as u8; 32]);
+            buffer.push(i, hash);
+        }
+
+        assert_eq!(buffer.entries.len(), BLOCKHASH_HISTORY_SIZE);
+
+        // Should only have the most recent 256 blocks
+        assert_eq!(
+            buffer.entries.front(),
+            Some(BlockHashEntry {
+                number: 44,
+                hash: B256::from([44; 32])
+            })
+            .as_ref()
+        ); // 300 - 256 = 44
+        assert_eq!(buffer.latest_block, 299);
+    }
+
+    #[test]
+    fn test_out_of_range_blocks() {
+        let mut buffer = BlockHashRingBuffer::default();
+
+        // Add block 100
+        buffer.push(100, B256::from([1u8; 32]));
+        buffer.push(101, B256::from([2u8; 32]));
+
+        // Block too far in the past (more than 256 blocks ago) should return None
+        // Current block is 101, so block 101 - 256 = -155 would be too old
+        // But let's test with a more realistic scenario
+        for i in 102..400u64 {
+            buffer.push(i, B256::from([(i % 256) as u8; 32]));
+        }
+
+        // Now current block is 399
+        // Block 100 should be too old (399 - 100 = 299 > 256)
+        assert_eq!(buffer.hash_by_number(100), None);
+
+        // Block 143 should still be accessible (399 - 143 = 256, which is the limit)
```

Comment:
Let's make a single comment saying it's not accessible because the limit should be less than 256. It's a strange comment saying it's accessible and then in the next line we say, oops just kidding.